driver-core/src/main/com/mongodb/ConnectionString.java

```diff
@@ -310,11 +310,11 @@ public ConnectionString(final String connectionString) {
                         + "If the username or password contains a colon (:) or an at-sign (@) then it must be urlencoded");
             }
             if (colonCount == 0) {
-                userName = urldecode(userInfo);
+                userName = urldecode(userInfo.replace("+", "%2B"));
```

Comment:
See the comment for why we're doing this.  The alternative is to copy URLDecoder and remove the special handling it has for the plus character.